cdap-storage-ext-spanner/src/main/java/io/cdap/cdap/storage/spanner/SpannerStructuredTableAdmin.java

```diff
@@ -0,0 +1,243 @@
+        @Override
+        public StructuredTableSchema load(StructuredTableId tableId) {
+          return loadSchema(tableId);
+        }
+      });
+  }
+
+  @Override
+  public void create(StructuredTableSpecification spec) throws IOException, TableAlreadyExistsException {
+    List<String> statements = new ArrayList<>();
+    statements.add(getCreateTableStatement(spec));
+
+    for (String idxColumn : spec.getIndexes()) {
+      statements.add(String.format("CREATE INDEX %s ON %s (%s)",
+                                   escapeName(getIndexName(spec.getTableId(), idxColumn)),
+                                   escapeName(spec.getTableId().getName()), escapeName(idxColumn)));
+    }
+
+    try {
+      Uninterruptibles.getUninterruptibly(adminClient.updateDatabaseDdl(databaseId.getInstanceId().getInstance(),
+                                                                        databaseId.getDatabase(), statements, null));
+    } catch (ExecutionException e) {
+      Throwable cause = e.getCause();
+      if (cause instanceof SpannerException
+        && ((SpannerException) cause).getErrorCode() == ErrorCode.FAILED_PRECONDITION) {
+        throw new TableAlreadyExistsException(spec.getTableId());
+      }
+      throw new IOException("Failed to create table in Spanner", cause);
+    }
+  }
+
+  @Override
+  public boolean exists(StructuredTableId tableId) {
+    try {
+      getSchema(tableId);
+      return true;
+    } catch (TableNotFoundException e) {
+      return false;
+    }
+  }
+
+  @Override
+  public StructuredTableSchema getSchema(StructuredTableId tableId) throws TableNotFoundException {
+    try {
+      return schemaCache.get(tableId);
+    } catch (ExecutionException | UncheckedExecutionException e) {
+      if (e.getCause() instanceof TableNotFoundException) {
+        throw (TableNotFoundException) e.getCause();
+      }
+      throw new RuntimeException("Failed to load table schema for " + tableId, e.getCause());
+    }
+  }
+
+  @Override
+  public void drop(StructuredTableId tableId) throws IOException {
+    List<String> ddlStatements = new ArrayList<>();
+
+    // Gather indexes of the table
+    Statement statement = Statement.newBuilder("SELECT index_name FROM information_schema.indexes " +
+                                                 "WHERE table_name = @table_name AND index_type = 'INDEX'")
+      .bind("table_name").to(tableId.getName())
+      .build();
+
+    try (ResultSet resultSet = databaseClient.singleUse().executeQuery(statement)) {
+      while (resultSet.next()) {
+        ddlStatements.add("DROP INDEX " + resultSet.getCurrentRowAsStruct().getString("index_name"));
+      }
+    }
+
+    try {
+      ddlStatements.add("DROP TABLE " + tableId.getName());
+      Uninterruptibles.getUninterruptibly(adminClient.updateDatabaseDdl(
+        databaseId.getInstanceId().getInstance(), databaseId.getDatabase(), ddlStatements, null));
```

Comment:
Remove table from schemaCache, otherwise it stays in cache forever, right?

Reply:
Good call.